### DISCUSSION TITLE
How to assign layers on separate pages using Simple Inkscape Scripting?

### DISCUSSION THREAD

**Discussion by MarySofy** [spakin/SimpInkScr#105] [Q&A] [answered]
Hi, i have another question. How do we assign a layer per page?

The code below will create two layers. The first layer is called 'red shape' & the second layer is called 'yellow shape". But it creates both the layers on the first page, and the second page is empty.

The result I would like to achieve is having the layer "red shape" on "red page" & the layer "yellow shape" on "yellow page" (...just to clarify, now both shapes are on the red page).

Is there a way to make each layer on separate pages ? Or perhaps just to save the shapes directly into a page, instead to a layer?

canvas.resize_by_name('A7')
c1 = circle((60, 60), 50, fill='magenta', stroke_width=4).to_path()
c2 = circle((90, 60), 50, fill='cyan', stroke_width=4).to_path()
color_list = ['red', 'none']
objs = apply_path_operation('division', [c1, c2])
for o, color in zip(objs, color_list):
    o.style(fill=color)
l2=layer ('red shape.svg')
l2.append(o)
pg = page('red page.svg')

canvas.resize_by_name('A7')
c1 = circle((60, 60), 50, fill='magenta', stroke_width=4).to_path()
c2 = circle((90, 60), 50, fill='cyan', stroke_width=4).to_path()
color_list = ['yellow', 'none']
objs = apply_path_operation('division', [c1, c2])
for o, color in zip(objs, color_list):
    o.style(fill=color)
l2=layer ('yellow shape.svg')
l2.append(o)
pg = page('yellow page.svg')

**Comment by spakin** [comment, marked as answer]
> Is there a way to make each layer on separate pages ? Or perhaps just to save the shapes directly into a page, instead to a layer?

Yes, but first let me explain the concept underlying layers and pages.  That should make the coding step more understandable.

Both layers and pages are Inkscape-specific concepts; they don't exist in plain SVG.  A layer is implemented as an SVG group.  Just like an ordinary group, its size is determined by its contents.  The only difference between a layer and a group is that a layer contains a couple Inkscape-specific attributes: `inkscape:groupmode` to indicate that the group should be interpreted as a layer and `inkscape:label` to specify the layer's name.  A page is simply a rectangular region of the canvas.  It is defined very much like an SVG rectangle but using an `<inkscape:page>` element instead of a `<rect>` element.  The Inkscape GUI pays attention to pages; other SVG viewers ignore them.

One corollary of the above is that a layer doesn't fundamentally exist on a particular page.  It's merely a grouping of objects that may all lie on a single page or that may lie on different pages.  A second corollary of the above is that placing an object on a particular page primarily is a matter of placing the object somewhere on the global canvas within the page's rectangular boundaries.

That said, here's a small piece of code that centers a red circle on the "red" page and a yellow circle on the "yellow" page:


```Python
canvas.resize_by_name('A7')

red_page = page('Red page')
rlft, rtop, rwd, rht = red_page.viewbox
red_circle = circle((rlft + rwd/2, rtop + rht/2), 50, fill='red', stroke_width=4)

yellow_page = page('Yellow page')
ylft, ytop, ywd, yht = yellow_page.viewbox
yellow_circle = circle((ylft + ywd/2, ytop + yht/2), 50, fill='yellow', stroke_width=4)
```

Notice how I use the page's `viewbox` property to acquire the page rectangle then use this to center the corresponding circle.  An alternative approach would be to translate each object by its target page's left and top coordinates.

It's easy enough to put each shape in its own layer, but remember that there's no formal association between layers and pages:

```Python
canvas.resize_by_name('A7')

red_page = page('Red page')
rlft, rtop, rwd, rht = red_page.viewbox
red_circle = circle((rlft + rwd/2, rtop + rht/2), 50, fill='red', stroke_width=4)
red_layer = layer('Red layer')
red_layer.append(red_circle)

yellow_page = page('Yellow page')
ylft, ytop, ywd, yht = yellow_page.viewbox
yellow_circle = circle((ylft + ywd/2, ytop + yht/2), 50, fill='yellow', stroke_width=4)
yellow_layer = layer('Yellow layer')
yellow_layer.append(yellow_circle)
```

I hope that helps you achieve what you want with your actual code.

**Reply by MarySofy** [reply]
Thank you so much for your help! I greatly appreciate it! I edited the code to what I need and now it works perfectly.
